include/pisa/payload_vector.hpp

```diff
@@ -0,0 +1,346 @@
+        {
+            Payload_Vector_Iterator next_iter{offset_iter, payload_iter};
+            ++(*this);
+            return next_iter;
+        }
+
+        constexpr auto operator--() -> Payload_Vector_Iterator & { return *this -= 1; }
+
+        [[nodiscard]]
+        constexpr auto operator--(int) -> Payload_Vector_Iterator
+        {
+            Payload_Vector_Iterator next_iter{offset_iter, payload_iter};
+            --(*this);
+            return next_iter;
+        }
+
+        [[nodiscard]]
+        constexpr auto operator+(size_type n) const -> Payload_Vector_Iterator
+        {
+            return {std::next(offset_iter, n),
+                    std::next(payload_iter, *std::next(offset_iter, n) - *offset_iter)};
+        }
+
+        [[nodiscard]]
+        constexpr auto operator+=(size_type n) -> Payload_Vector_Iterator &
+        {
+            std::advance(payload_iter, *std::next(offset_iter, n) - *offset_iter);
+            std::advance(offset_iter, n);
+            return *this;
+        }
+
+        [[nodiscard]]
+        constexpr auto operator-(size_type n) const -> Payload_Vector_Iterator
+        {
+            return {std::prev(offset_iter, n),
+                    std::prev(payload_iter, *offset_iter - *std::prev(offset_iter, n))};
+        }
+
+        [[nodiscard]]
+        constexpr auto operator-=(size_type n) -> Payload_Vector_Iterator &
+        {
+            return *this = *this - n;
+        }
+
+        [[nodiscard]]
+        constexpr auto operator-(Payload_Vector_Iterator const& other) -> difference_type
+        {
+            return offset_iter - other.offset_iter;
+        }
+
+        [[nodiscard]]
+        constexpr auto operator*() -> value_type
+        {
+            return value_type(reinterpret_cast<char const *>(&*payload_iter),
+                              *std::next(offset_iter) - *offset_iter);
+        }
+
+        [[nodiscard]] constexpr auto operator==(Payload_Vector_Iterator const &other) const -> bool
+        {
+            return offset_iter == other.offset_iter;
+        }
+
+        [[nodiscard]] constexpr auto operator!=(Payload_Vector_Iterator const &other) const -> bool
+        {
+            return offset_iter != other.offset_iter;
+        }
+    };
+
+    template <typename... Ts>
+    struct all_pod;
+
+    template <typename Head, typename... Tail>
+    struct all_pod<Head, Tail...> {
+        static const bool value = std::is_pod_v<Head> && all_pod<Tail...>::value;
+    };
+
+    template <typename T>
+    struct all_pod<T> {
+        static const bool value = std::is_pod_v<T>;
+    };
+
+    template <typename... Ts>
+    struct sizeofs;
+
+    template <typename Head, typename... Tail>
+    struct sizeofs<Head, Tail...> {
+        static const size_t value = sizeof(Head) + sizeofs<Tail...>::value;
+    };
+
+    template <typename T>
+    struct sizeofs<T> {
+        static const size_t value = sizeof(T);
+    };
+
+    struct test {
```

Comment:
Should this be here?

Reply:
Yes, I want to have a compile-time check for this, to make sure that these work as expected. It's in `detail` so I figure it's as good a place as any.